### ISSUE TITLE
Connection via proxy is not working

### ISSUE THREAD

**Issue by thomaslevesque** [kerryjiang/WebSocket4Net#34]
I'm trying to connect to websocket via a Squid proxy. I'm passing the IP address and port via the `httpConnectProxy` parameter of the `WebSocket` constructor.

If I try to connect to a ws:// URL, I get a 400 error from the proxy. With a wss:// URL, I get an SSL error ("The handshake failed due to an unexpected packet format").

The same scenarios (normal and secure connection) work fine with `System.Net.WebSockets.ClientWebSocket`. So I used Wireshark to compare the behavior of both implementation, and I realized that `System.Net.WebSockets.ClientWebSocket` is using HTTP CONNECT tunelling, whereas WebSocket4Net is sending the handshake request directly to the proxy. This behavior means that :
- if we're connecting via HTTPS, the proxy has to be HTTPS as well, otherwise the SSL handshake doesn't work
- if the proxy doesn't understand the websocket URL scheme, it just rejects the request with a 400 error

IMO WebSocket4Net should use HTTP CONNECT tunelling, it would work in much more situations.

Also, I see there's a `Proxy` property in addition to the `httpConnectProxy` parameter. What's the difference? What can I set the property to ? I don't see any implementation of `IProxyConnector` in the library.

**Comment by RobertoSchneiders** [comment]
Have you tried [this](http://stackoverflow.com/a/23417094/1847499) @thomaslevesque?

**Comment by thomaslevesque** [comment]
[**@RobertoSchneiders**](https://github.com/RobertoSchneiders) commented on [22 sept. 2015 15:52 UTC+2](https://github.com/kerryjiang/WebSocket4Net/issues/34#issuecomment-142295414 "2015-09-22T13:52:06Z - Replied by Github Comment Enhancer"):
> Have you tried [this](http://stackoverflow.com/a/23417094/1847499) [@thomaslevesque](https://github.com/thomaslevesque)?

No, I haven't. I'll give it a try. Thanks!

**Comment by kerryjiang** [comment]
@thomaslevesque there is assembly in SuperSocket.ClientEgnine called SuperSocket.ClientEngine.Proxy. All implementations of IProxyConnector locate in it.

**Comment by thomaslevesque** [comment]
@kerryjiang, where is that assembly? It doesn't seem to be in the NuGet package.

Anyway, I don't think I can use it directly, because of the way SuperSocket is ILMerged into WebSocket4Net. SuperSocket.ClientEngine.Proxy references the `IProxyConnector` from SuperSocket.ClientEngine, not the one from WebSocket4Net...

**Comment by kerryjiang** [comment]
Try download it from here:
https://github.com/kerryjiang/WebSocket4Net/tree/master/Reference/SuperSocket.ClientEngine/Net40/Release

**Comment by thomaslevesque** [comment]
@kerryjiang , thanks, but it still has the assembly identity problem I mentioned... The only option seems to be to copy the `HttpConnectProxy` code into my project, as suggested in the StackOverflow answer linked by @RobertoSchneiders 

Also, `HttpConnectProxy` doesn't seem to support proxy authentication (or am I missing something?)

**Comment by RobertoSchneiders** [comment]
I think it is possible to use proxy authentication adding the user and password on the url, something like this: 
```
http://username:password@proxy_url:port
```

I'm not sure

**Comment by thomaslevesque** [comment]
@RobertoSchneiders, no, because you don't pass a URI, you pass a `EndPoint` (`DnsEndPoint` or `IPEndPoint`). I think I can pass the credentials in a `Proxy-Authorization` HTTP header, but right now I'm hitting other issues related to the buffer size...

**Comment by kerryjiang** [comment]
There is another parameter for http header items.

**Comment by thomaslevesque** [comment]
[**@kerryjiang**](https://github.com/kerryjiang) commented on [22 sept. 2015 18:44 UTC+2](https://github.com/kerryjiang/WebSocket4Net/issues/34#issuecomment-142344722 "2015-09-22T16:44:19Z - Replied by Github Comment Enhancer"):
> There is another parameter for http header items.

Yes, I'm already using it ;)

**Comment by thomaslevesque** [comment]
OK, I eventually managed to connect, but I had to make several changes to `HttpConnectProxy`:
- add an `Authorization` property to be able to specify the proxy credentials
- add the `Proxy-Authorization` header if the `Authorization` property is set
- change the response header validation to accept "HTTP/1.0" as well as "HTTP/1.1" (the proxy I'm using for my tests is an old Squid proxy and always returns "HTTP/1.0")

But it's still not robust enough; if the proxy returns any status other than 2xx, `HttpConnectProxy` just throws an error saying "the proxy server refused the connection", with no details at all. So I can't even detect if the credentials are incorrect...

**Comment by RobertoSchneiders** [comment]
@thomaslevesque Can you share your code? I'll suggest to you send a Pull Request with your work  in progress. 

It probably will make easier to understand the problems and help you. I don't know if I will be able to help, but I can try. I also need this proxy support.

**Comment by thomaslevesque** [comment]
@RobertoSchneiders, here are my changes: https://github.com/kerryjiang/SuperSocket.ClientEngine/commit/18109a20bdbe9d83709e64b7b3c9e650e4af8a94

I don't want to send a pull request right now, because it's more a POC than anything else at this point. The value of the `Proxy-Authorization` header has to be provided manually, which isn't very convenient. Also, I'd like to add some specific handling for HTTP status 407 (Proxy authentication required), e.g. extract the authentication challenge from the `Proxy-Authenticate` header and include it in the exception.

**Comment by thomaslevesque** [comment]
@kerryjiang, it seems I'm unable to receive the response if it contains a body, I get the error "protocol error: more data has been received". It happens if the server returns HTTP status 407 for instance; why does it matter if there is a body after the headers ? It should just be ignored... I think this test should be removed:

            if (e.Offset + e.BytesTransferred > responseLength + m_LineSeparator.Length)
            {
                OnException("protocol error: more data has been received");
                return;
            }

Do you see any reason to keep it?

**Comment by osullivj** [comment]
Thanks @thomaslevesque for the auth patch and @RobertoSchneiders for the proxy fix. I've applied both in my Excel addin - SSAddin - which @govert's ExcelDNA to push websocket sourced ticking data into Excel.

**Comment by sohaimeng** [comment]
Followed this http://stackoverflow.com/questions/23024121/how-to-use-proxies-with-the-websocket4net-library and it hit exception that the remoteEndPoint is null in Connect method in SuperSocket.ClientEngine.Proxy.

**Comment by gkalyanxo** [comment]
@sohaimeng, I am facing the same issue. Were you able to figure out how to fix it? Thanks.

**Comment by scolestock** [comment]
I could certainly be wrong, but I believe that the changes introduced in [this commit](https://github.com/kerryjiang/SuperSocket.ClientEngine/commit/4ae90a56558fa5b042fde9238a31b8e98e536a4d) of SuperSocket.ClientEngine likely caused proxies to stop working correctly.  They work for us back in the 0.14.1 version (c54a0cf).  The refactoring in WebSocket.cs between these two versions to accommodate the change in AsyncTcpSession may have caused this to break.

**Comment by kerryjiang** [comment]
OK, I'll review the code.

I am really sorry about my late responses about the issues of my open source projects.
I just moved to US a couple of months ago, so I am very busy on my work and my personal affairs.

**Comment by sohaimeng** [comment]
The latest beta version 0.15.0-beta6 solved remoteEndPoint is null problem. 

I wonder if "the proxy server refused the connection" & "proxy error ---> System.Exception: receive buffer size has been exceeded" had been solved.

I tested 3 approaches where the 1st works fine but others not working :
1. Use fiddler as proxy (without proxy credential) and it works fine, able to see websocket stream in fiddler.
2. Use fiddler as proxy (with proxy credential) and get "the proxy server refused the connection". I believe i used the websocket correctly by passing
`string proxyAuthInfo = String.Join(":", proxyUsername,proxyPassword);`
`proxyAuthInfo = Convert.ToBase64String(Encoding.Default.GetBytes(proxyAuthInfo));`
`customHeaderList = new List<KeyValuePair<string, string>> {  new KeyValuePair<string, string>("Authorization", "Basic " + proxyAuthInfo)};`
3. Use squid as proxy (with proxy credential) and get "proxy error ---> System.Exception: receive buffer size has been exceeded" had been solved".
